typing/ctype.ml

```diff
@@ -890,30 +890,38 @@ let rec lower_contravariant env var_level visited contra ty =
   in
   if must_visit then begin
     Hashtbl.add visited ty.id contra;
-    let generalize_rec = lower_contravariant env var_level visited in
+    let lower_rec = lower_contravariant env var_level visited in
     match ty.desc with
       Tvar _ -> if contra then set_level ty var_level
-    | Tconstr (path, tyl, abbrev) ->
+    | Tconstr (_, [], _) -> ()
+    | Tconstr (path, tyl, _abbrev) ->
         let variance =
           try (Env.find_type path env).type_variance
           with Not_found ->
             (* See testsuite/tests/typing-missing-cmi-2 for an example *)
             List.map (fun _ -> Variance.may_inv) tyl
         in
-        abbrev := Mnil;
-        List.iter2
-          (fun v t ->
-            if Variance.(mem May_weak v)
-            then generalize_rec true t
-            else generalize_rec contra t)
-          variance tyl
+        if List.for_all ((=) Variance.null) variance then () else
+        begin try
+          let ty = !forward_try_expand_once env ty in
+          lower_rec contra ty
+        with Cannot_expand ->
```

Comment:
This would be more readable with

```ocaml
match !forward_try_expand_once ty with
| ty -> lower_rec contra ty
| exception Cannot_expand ->
  ...
```

which has more localized exception handling (and a tail-call, which is nice).

Reply:
Indeed, I even remember noting to myself that this try-with was not tail-recursive.
Funny how long it takes for new language features to become automatic.